project/FormatStagedScalaFilesPlugin.scala

```diff
@@ -0,0 +1,76 @@
+import better.files.File
+import org.scalafmt.sbt.ScalafmtPlugin
+import sbt.Keys.streams
+import sbt.nio.Keys.{ReloadOnSourceChanges, onChangedBuildSource}
+import sbt.{Compile, Global, Setting, taskKey}
+import utils.Step
+
+object FormatStagedScalaFilesPlugin extends sbt.AutoPlugin {
+  override def trigger = allRequirements
+
+  object autoImport {
+    val formatStagedScalaFiles = taskKey[Unit]("Format staged Scala files")
+  }
+
+  import autoImport._
+
+  override def projectSettings = Seq(
+    formatStagedScalaFiles in Global := {
+      formatStagedScalaFilesOnly().value
+    }
+  )
+
+  override def globalSettings: Seq[Setting[_]] = Seq(
+    Global / onChangedBuildSource := ReloadOnSourceChanges,
+  )
+
+  private def formatStagedScalaFilesOnly() = {
+    val result = for {
+      stagedFiles <- getStagedScalaFiles()
+      _ <-
+        if (stagedFiles.nonEmpty) {
+          for {
+            _ <- callFormatFiles(stagedFiles)
+            _ <- addToGitAllStagedFilesOnceAgain(stagedFiles)
+          } yield ()
+        } else {
+          Step.taskUnit
+        }
+    } yield ()
+    result.runThrowing
+  }
+
+  private def getStagedScalaFiles() = Step.deferredTask {
+    os
+      .proc("git", "diff", "--cached", "--name-only", "--diff-filter=ACM")
+      .call()
+      .out
+      .lines()
+      .filter(f => f.endsWith(".scala") || f.endsWith(".sbt"))
+      .toList
+  }
+
+  private def callFormatFiles(files: List[String]) = {
+    val userDir       = getUserDir()
+    val fullPathFiles = files.map(f => s"${(userDir / f).path}")
+    for {
+      _ <- Step.task {
+        streams.map(_.log.info("Formatting backend files ..."))
+      }
+      _ <- Step.task {
+        (Compile / ScalafmtPlugin.autoImport.scalafmtOnly).toTask(s" ${fullPathFiles.mkString(" ")}")
```

Comment:
This will fail if project is checked out eg. at `/home/My Projects` (or any directory containing space) - paths should be quoted and escaped.

Reply:
... or is it sbt-scalafmt limitation? Maybe let's throw an error if we see that `userDir` contains space character?

Reply:
I've tested and it works now.